plugins/soundsourcem4a/soundsourcem4a.cpp

```diff
@@ -561,33 +556,48 @@ SINT SoundSourceM4A::readSampleFrames(
         const SINT numberOfSamplesDecoded = decFrameInfo.samples;
         DEBUG_ASSERT(numberOfSamplesDecoded <= decodeBufferCapacity);
         SINT numberOfSamplesRead;
-        if (pDecodeBuffer == pSampleBuffer) {
-            numberOfSamplesRead = math_min(numberOfSamplesDecoded, numberOfSamplesRemaining);
-            pSampleBuffer += numberOfSamplesRead;
+        if (writableSampleFrames.sampleBuffer().data() &&
+                (pDecodeBuffer == writableSampleFrames.sampleBuffer().data(outputSampleOffset))) {
+            // Decoded in-place
+            DEBUG_ASSERT(numberOfSamplesDecoded <= numberOfSamplesRemaining);
+            numberOfSamplesRead = numberOfSamplesDecoded;
+            outputSampleOffset += numberOfSamplesRead;
         } else {
-            m_sampleBuffer.readFromTail(decodeBufferCapacity - numberOfSamplesDecoded);
-            const SampleBuffer::ReadableChunk readableChunk(
-                    m_sampleBuffer.readFromHead(numberOfSamplesRemaining));
-            numberOfSamplesRead = readableChunk.size();
-            if (pSampleBuffer) {
-                SampleUtil::copy(pSampleBuffer, readableChunk.data(), numberOfSamplesRead);
-                pSampleBuffer += numberOfSamplesRead;
+            // Decoded into temporary buffer
+            DEBUG_ASSERT(numberOfSamplesDecoded <= decodeBufferCapacity);
+            // Adjust the size of the buffer to the samples that
+            // have actually been decoded
+            m_sampleBuffer.readFifo(decodeBufferCapacity - numberOfSamplesDecoded);
```

Comment:
The function parameter is provided by the caller and we are only allowed to write the requested samples in there. But the decoder often produces more samples on each invocation that need to be stored temporarily. That's why I even changed the class name to ReadAheadSampleBuffer, because it's main purpose of this FIFO is to buffer superfluous sample data until the next round.

Reply:
Ah sorry, I have misread the code. Now on the second read everything makes sense and it really looks rock solid. The only issue is that you need to wrap your head around it, which is quite hard because all that is not done the usually way as you might expect. ... continued inline ...

Reply:
It looks like we never actually read from the back of the buffer, so you may remove the current readFifo function and replace it with a shrink function.

Reply:
It is also hard to understand that we first write the buffer above, out of sight on on screen page and than adjust the buffer here, under an if condition. Cant we do it step by step just after writing, or do I miss something. 

(Sorry for all these stupid questions, if they are stupid .. :-) )